src/analytics/src/main/scala/osmesa/analytics/stats/CalculateStats.scala

```diff
@@ -0,0 +1,365 @@
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql._
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+import org.apache.spark.sql.functions._
+import vectorpipe._
+
+import java.math.BigDecimal
+import java.time.Instant
+import scala.collection.mutable
+import scala.util.{Try, Success, Failure}
+
+object CalculateStats {
+  private implicit def changeTopicEncoder: Encoder[Array[StatTopic]] = ExpressionEncoder()
+
+  case class Options(
+    changesetPartitionCount: Int = 1000,
+    wayPartitionCount: Int = 1000,
+    nodePartitionCount: Int = 10000
+  )
+
+  object Options {
+    def DEFAULT = Options()
+  }
+
+  def statTopics(col: Column): TypedColumn[Any, Array[StatTopic]] =
+    udf[Array[StatTopic], Map[String, String]](StatTopics.tagsToTopics).
+      apply(col).
+      as[Array[StatTopic]]
+
+  def computeChangesetStats(history: DataFrame, changesets: DataFrame, options: Options = Options.DEFAULT)(implicit ss: SparkSession): RDD[(Long, ChangesetStats)] = {
+    import ss.implicits._
+
+    val changesetPartitioner = new HashPartitioner(options.changesetPartitionCount)
+    val wayPartitioner = new HashPartitioner(options.wayPartitionCount)
+    val nodePartitioner = new HashPartitioner(options.nodePartitionCount)
+
+    def generateUpstreamTopics[T <: OsmId](rdd: RDD[(Long, Iterable[(T, Long, Set[StatTopic])])]): RDD[(Long, List[(T, UpstreamTopics)])] =
+      rdd.
+        mapValues { idsAndChangesets =>
+          (idsAndChangesets.foldLeft(Map[T, Map[Long, Set[StatTopic]]]()) { case (acc, (osmId, changeset, statTopics)) =>
+            acc.get(osmId) match {
+              case Some(m) => acc + (osmId -> (m + (changeset ->  statTopics)))
+              case _ => acc + (osmId -> Map(changeset -> statTopics))
+            }
+          }).toList.map { case (k, v) => (k, new UpstreamTopics(v)) }
+        }
+
+
+    // Create the base set of ChangesetStats we'll be joining against
+    val initialChangesetStats =
+      changesets.
+        select($"id", $"uid", $"user", $"created_at", $"closed_at", $"tags", hashtags($"tags").alias("hashtags")).
+        map { row =>
+          val changeset = row.getAs[Long]("id")
+          val userId = row.getAs[Long]("uid")
+          val userName = row.getAs[String]("user")
+          val createdAt = row.getAs[java.sql.Timestamp]("created_at")
+          val closedAt = row.getAs[java.sql.Timestamp]("closed_at")
+          val tags = row.getAs[Map[String, String]]("tags")
+          val editor = tags.get("created_by")
+
+          val hashtags = row.getAs[Seq[String]]("hashtags").toList
+
+          (changeset, userId, userName, createdAt, closedAt, editor, hashtags)
+        }.
+        rdd.
+        map { case (changeset, userId, userName, createdAt, closedAt, editor, hashtags) =>
+          val stats =
+            ChangesetStats(
+              changeset,
+              userId,
+              userName,
+              createdAt.toInstant,
+              closedAt.toInstant,
```

Comment:
I've noticed the use of `Instant`s elsewhere; is there a particular reason to use them over `Timestamp` (other than `java.time` seeming more general than `java.sql`.

I ask because converting these back into `DataFrame`s (for the purpose of writing them out) fails due to a lack of `java.time.Instant` encoder.

```scala
val statsDF = changesetStats.values.toDF("changeset", "uid", "username", "created_at", "closed_at", "editor", "hashtags", "roads_added", "roads_modified", "buildings_added", "buildings_modified", "waterways_added", "waterways_modified", "pois_added", "pois_modified", "km_road_added", "km_road_modified", "km_waterway_added", "km_waterway_modified", "countries")
statDF.repartition(1).write.format("json").save("stats")
```

(The above works after leaving them as `Timestamp`s.)

Reply:
The domain models that Nathan was working on had Instants. Definitely not tied to the type; some of that might be fleshed out when we combine this PR with the query server PR to get on the same case classes.

Reply:
(Don't consider this a problem; I've figured out how to map them back to `Timestamp`s before converting to `DataFrame`s.)

Reply:
@mojodna Can you describe how to convert Instant fields back to Dataframes

Reply:
First, convert them to `Timestamp`s (since Spark knows how to serialize them):

```scala
import java.time.Instant
import java.sql.Timestamp

new Timestamp(Instant.now.toEpochMilli)
```

Then, `toDF` should just work.